api/org/commcare/api/persistence/SqlIndexedStorageUtility.java

```diff
@@ -0,0 +1,418 @@
+     */
+    public int add(T e) throws StorageFullException {
+        this.write(e);
+        return 1;
+    }
+
+    /* (non-Javadoc)
+     * @see org.javarosa.core.services.storage.IStorageUtility#close()
+     */
+    public void close() {
+        // TODO Auto-generated method stub
+
+    }
+
+    /* (non-Javadoc)
+     * @see org.javarosa.core.services.storage.IStorageUtility#destroy()
+     */
+    public void destroy() {
+        // TODO Auto-generated method stub
+
+    }
+
+    /* (non-Javadoc)
+     * @see org.javarosa.core.services.storage.IStorageUtility#exists(int)
+     */
+    public boolean exists(int id) {
+        PreparedStatement preparedStatement = null;
+        Connection c = null;
+        try {
+            c = getConnection();
+            preparedStatement = SqlHelper.prepareIdSelectStatement(c, this.tableName, id);
+            if (preparedStatement == null) {
+                return false;
+            }
+            ResultSet rs = preparedStatement.executeQuery();
+            if (rs.next()) {
+                return true;
+            }
+        } catch (Exception e) {
+            System.out.println("SqlIndexedStorageUtility readBytes exception: " + e);
+        } finally {
+            try {
+                if (c != null) {
+                    c.close();
+                }
+                if (preparedStatement != null) {
+                    preparedStatement.close();
+                }
+            } catch (SQLException e) {
+                e.printStackTrace();
+            }
+        }
+        return false;
+    }
+
+    /* (non-Javadoc)
+     * @see org.javarosa.core.services.storage.IStorageUtility#getAccessLock()
+     */
+    public Object getAccessLock() {
+        // TODO Auto-generated method stub
+        return null;
+    }
+
+    @Override
+    public int getNumRecords() {
+        PreparedStatement preparedStatement = null;
+        Connection c = null;
+        try {
+            c = getConnection();
+            String sqlQuery = "SELECT COUNT (*) FROM " + this.tableName + ";";
+            preparedStatement = c.prepareStatement(sqlQuery);
+            ResultSet rs = preparedStatement.executeQuery();
+            return rs.getInt(1);
+        } catch (Exception e) {
+            System.out.println("SqlIndexedStorageUtility readBytes exception: " + e);
+        } finally {
+            try {
+                if (c != null) {
+                    c.close();
+                }
+                if (preparedStatement != null) {
+                    preparedStatement.close();
+                }
+            } catch (SQLException e) {
+                e.printStackTrace();
+            }
+        }
+        return -1;
+    }
+
+    /* (non-Javadoc)
+     * @see org.javarosa.core.services.storage.IStorageUtility#getRecordSize(int)
+     */
+    public int getRecordSize(int id) {
+        //serialize and test blah blah.
```

Comment:
More importantly, we should consider, at some point, refactoring the `IStorageUtility`/`IStorageUtilityIndexed` interface such that we can isolate methods that are only used by J2ME code. I dislike seeing empty/non-functional method declarations just because we didn't manage our interface abstractions correctly